compiler-core/src/build/project_compiler.rs

```diff
@@ -562,13 +566,31 @@ where
         };
 
         // Compile project to Erlang or JavaScript source code
-        compiler.compile(
+        let outcome = compiler.compile(
             &mut self.warnings,
             &mut self.importable_modules,
             &mut self.defined_modules,
             &mut self.stale_modules,
+            &self.incomplete_modules,
             self.telemetry.as_ref(),
-        )
+        );
+        let outcome = match outcome {
+            Outcome::Ok(modules) => {
+                // On successful compilation we remove the module from the incomplete set
+                modules.iter().for_each(|m| {
+                    let _ = self.incomplete_modules.remove(&m.name);
+                });
+                Outcome::Ok(modules)
+            }
+            Outcome::PartialFailure(modules, errors) => {
+                // On partial compilation failure we add all the module to the incomplete set
```

Comment:
It's not just these that are incomplete, this also contains modules that have compiles successfully. If we're going to track this we need to make it accurate.

This data structure is being manipulated within the compiler, but then for this part we do it here. Perhaps it should be done inside instead? Though there's not a clear point where it has information about which is the one that failed and as such is incomplete.

Reply:
Good point, yeah I can move the insert to at the point of failure instead (I think the set itself still needs to be stored here so as to make sure the lsp maintains it/doesn't throw it out but the inserts/clears can definitely be moved)